### PR TITLE
NXP-29585: Add Bulk command queryLimit

### PR DESCRIPTION
**Pull request by bdelbosc**
A default limit can be set on action and overriden when building command,
a flag in the status indicate if the limit has been reached for a command.

### CODE REVIEW

**Review by kevinleturc — DISMISSED** [review]
I think your commit _NXP-29585: Bump default partitions for async processing_ doesn't have the right NXP, shouldn't it be _NXP-29677_?

**Review by kevinleturc — APPROVED** [review]

File: modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java [hunk 1]
```diff
@@ -111,6 +112,9 @@
     @Nullable
     protected Long processingDurationMillis;
 
+    @AvroDefault("false")
+    protected boolean limitedQuery;
```
Comment (efge) [inline modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java:116]: It's not clear to me what his means when it's true. Is it that there was a limit passed when doing the query? Or that the query is incomplete in its results returned.
A useful flag would be more about "limitReached", which I would call "resultsTruncated" or "resultsIncomplete", or "moreResults".

File: modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java [hunk 2]
```diff
@@ -169,21 +176,33 @@ protected void processRecord(ComputationContext context, Record record) {
...
                     List<String> docIds = scroll.next();
-                    documentIds.addAll(docIds);
+                    int scrollCount = docIds.size();
+                    if (documentCount + scrollCount < queryLimit) {
+                        documentIds.addAll(docIds);
+                    } else {
+                        scrollCount = Math.toIntExact(queryLimit - documentCount);
+                        documentIds.addAll(docIds.subList(0, scrollCount));
+                        limitReached = true;
+                    }
```
Comment (efge) [inline modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java:186]: What about if we get exactly to the queryLimit value? Then we flag limitReached=true but there's no additional documents. The use of this for paging is not clear.

File: modules/core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/action/TestWordCountAction.java [hunk 3]
```diff
@@ -70,9 +71,57 @@ public void testWordCount() throws Exception {
...
+                .queryUnlimited()
+                .build();
+        commandId = bulkService.submit(command);
+        assertTrue(bulkService.await(commandId, Duration.ofSeconds(20)));
+        status = bulkService.getStatus(commandId);
+        assertEquals(COMPLETED, status.getState());
+        assertFalse(status.isLimitedQuery());
+        long resultUnlimited = (Integer) status.getResult().get("wordCount");
+        assertEquals(wordCount, resultUnlimited);
+    }
```
Comment (efge) [inline modules/core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/action/TestWordCountAction.java:123]: Please add tests for the behavior where the `.queryLimit()` is wordCount-1, exactly wordCount, and wordCount+1 to test edge cases.

**Review by efge — CHANGES_REQUESTED** [review]

File: modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java (hunk 1 above)
Comment (bdelbosc) [reply modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java:116]: limitedQuery is true if the query has a limit and it has been reached which makes sense in the context of the status, the query itself and its attributes (like the limit) are part of the command. limitReached or queryLimitReached could be an alternative.

File: modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java (hunk 2 above)
Comment (bdelbosc) [reply modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java:186]: the limit should be taken as LIMIT in SQL, the bulk service is not done for paging and we don't have to go into an off-by-one problem here.

File: modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java (hunk 1 above)
Comment (bdelbosc) [reply modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkStatus.java:116]: will use queryLimitReached

**Review by efge — APPROVED** [review]

**Review by kevinleturc — APPROVED** [review]
